example-app/src/main/java/com/adyen/checkout/example/ui/card/CardFragment.kt

```diff
@@ -0,0 +1,99 @@
+import androidx.fragment.app.viewModels
+import androidx.lifecycle.lifecycleScope
+import com.adyen.checkout.card.CardComponent
+import com.adyen.checkout.components.model.paymentmethods.PaymentMethod
+import com.adyen.checkout.example.databinding.FragmentCardBinding
+import com.adyen.checkout.example.ui.configuration.CheckoutConfigurationProvider
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import dagger.hilt.android.AndroidEntryPoint
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class CardFragment : BottomSheetDialogFragment() {
+
+    @Inject
+    internal lateinit var checkoutConfigurationProvider: CheckoutConfigurationProvider
+
+    private var _binding: FragmentCardBinding? = null
+    private val binding: FragmentCardBinding get() = requireNotNull(_binding)
+
+    private val cardViewModel: CardViewModel by viewModels()
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        _binding = FragmentCardBinding.inflate(inflater, container, false)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.payButton.setOnClickListener { cardViewModel.onPayClick() }
+
+        lifecycleScope.launchWhenStarted {
+            launch { cardViewModel.cardViewState.collect(::onCardViewState) }
+            launch { cardViewModel.paymentResult.collect(::onPaymentResult) }
+        }
+    }
+
+    private fun onCardViewState(cardViewState: CardViewState) {
+        when (cardViewState) {
+            CardViewState.Loading -> {
+                binding.progressIndicator.isVisible = true
+                binding.cardContainer.isVisible = false
+                binding.errorView.isVisible = false
+            }
+            is CardViewState.Data -> {
+                binding.cardContainer.isVisible = true
+                binding.progressIndicator.isVisible = false
+                binding.errorView.isVisible = false
+
+                setupCardView(cardViewState.paymentMethod)
+            }
+            CardViewState.Invalid -> {
+                binding.cardView.highlightValidationErrors()
+            }
+            CardViewState.Error -> {
+                binding.errorView.isVisible = true
+                binding.progressIndicator.isVisible = false
+                binding.cardContainer.isVisible = false
+            }
+        }
+    }
+
+    private fun setupCardView(paymentMethod: PaymentMethod) {
+        val cardComponent = CardComponent.PROVIDER.get(
+            this,
+            paymentMethod,
+            checkoutConfigurationProvider.getCardConfiguration()
+        )
+
+        binding.cardView.attach(cardComponent, viewLifecycleOwner)
+
+        cardComponent.observe(viewLifecycleOwner, cardViewModel::onCardComponentState)
```

Comment:
Maybe we should also add `observeError` to illustrate a possible failure scenario.

Reply:
Good one! Wasn't aware of this separate error stream. Didn't see it in the docs as well, should it be mentioned there? https://docs.adyen.com/online-payments/android/components